### PR TITLE
auth: return 401 for invalid auth headers on DotCom

### PR DESCRIPTION
**Pull request by evict**
In our current setup, when a request with incorrect authentication headers reaches DotCom, it's mistakenly treated with a 429 status code due to how we handle rate limits for anonymous requests. DotCom mode-specific behaviour was identified: the system keeps forwarding the wrong header until it reaches the [middleware that checks if we allow anonymous access](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/auth/non_public.go?L23-37).
This middleware passes the check as it's configured to identify if we are operating in [DotCom mode](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/conf/auth.go?L28-33).

I propose to check if we are in DotCom mode when we have an invalid header, to prevent us from treating users that supply (invalid) authorization headers as anonymous users. 

See Slack for context: https://sourcegraph.slack.com/archives/C05Q75F9ES3/p1695786335453289.

## Test plan
Tested locally with an instance in DotCom mode.

### CODE REVIEW

**Comment by evict** [comment]
@jhchabran do we still need this auth header logging code?

**Review by unknwon — COMMENTED** [review]
TBH I was expecting to see 

```go
http.Error(w, "Invalid Authorization header.", http.StatusUnauthorized)
```

On line 84, and remove 

```go
next.ServeHTTP(w, r)
```

😂

**Comment by evict** [comment]
> TBH I was expecting to see 
> 
> 
> 
> ```go
> 
> http.Error(w, "Invalid Authorization header.", http.StatusUnauthorized)
> 
> ```
> 
> 
> 
> On line 84, and remove 
> 
> 
> 
> ```go
> 
> next.ServeHTTP(w, r)
> 
> ```
> 
> 
> 
> 😂 

I want to keep the functionality intact as it had been added for customer instances AFAIK. But it's useless for DotCom, hence the added check. It will throw invalid token errors for DotCom users.

**Review by jhchabran — CHANGES_REQUESTED** [review]
@evict the PR fixes the problem but does so in a way that further confuses readers. 

Thanks btw for fixing the comment with sha256/md5sum 🙏 

- We don't want to respond with a 429 to folks sending us incorrect authentication headers.
- This [slack message](https://sourcegraph.slack.com/archives/C05Q75F9ES3/p1695920754119599?thread_ts=1695786335.453289&cid=C05Q75F9ES3) says that we do return this 429 because anonymous requests end up with a rate limit quota of zero.
- Some users end up with requests with no authentication, which the current code incorrectly treats as having an unrecognized auth scheme, whereas it's actually none. 
- The fix we have here adds an additional test to prevent requests from entering this conditional branch if we're on dotcom, sending users back to the normal 401 route. 

So basically, we're compensating the fact that the code is not handling properly a case by adding another condition to force the 401 to happen. 

1/ If that's a hotfix, can you please add a comment to explain the reasoning? 

The PR description itself is very short in context and the Slack link will expire in 3 months, losing all context by then. Whoever stumbles across this will scratch their head for a moment before understanding what we're trying to achieve. 

2/ Can we make sure we follow up with a proper fix that makes it so that not passing any authentication at all ends up with a 401 regardless if it's dotcom or not? It's not important if that's you or someone else doing it.

**Comment by evict** [comment]
> The PR description itself is very short in context and the Slack link will expire in 3 months, losing all context by then. Whoever stumbles across this will scratch their head for a moment before understanding what we're trying to achieve.

Thanks you for pointing that out. I've updated the PR description to be more complete.

> * The fix we have here adds an additional test to prevent requests from entering this conditional branch if we're on dotcom, sending users back to the normal 401 route.

DotCom is the only instance where this behaviour would occur. The issue is that we keep passing along the wrong header until we [reach the middleware that checks if we allow anonymous access](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/auth/non_public.go?L23-37).  This middleware passes as, [it checks if we are in DotCom mode](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/conf/auth.go?L28-33).

To me it makes sense that if we are on DotCom mode, and we have an invalid header, we should not continue to the next middleware and actually return the error. There's no other middleware that would pick this up as AFAIK we don't support other `Authorization` headers for users.

> 2/ Can we make sure we follow up with a proper fix that makes it so that not passing any authentication at all ends up with a 401 regardless if it's dotcom or not? It's not important if that's you or someone else doing it.

To use DotCom publicly without authentication, we need to be able to make requests to  `/.api/grapqhl` without any authorization. DotCom is really an edge case here.

**Comment by jhchabran** [comment]
What I was trying to say, is that https://github.com/sourcegraph/sourcegraph/blob/949d26d267f692b0395703c34fb66cf61634d44e/internal/authz/header.go#L48-L50 should not return a `errUnrecognizedScheme` if there are no headers for auth at all. 

This way you can handle the cases more clearly: 

- dotcom
  - unrecognized header -> 401 
  - no auth -> pass through and let the next middleware handle it 

- not dotcom
  - unrecognized header -> 400 
  - no auth -> 401 

> To me it makes sense that if we are on DotCom mode, and we have an invalid header, we should not continue to the next middleware and actually return the error. There's no other middleware that would pick this up as AFAIK we don't support other Authorization headers for users.

Agreed. Again, the present fix is fine as a hotfix, I'm pointing out that this could be simplified and easier to read :)

**Review by jhchabran — APPROVED** [review]
